clang/lib/Frontend/CompilerInvocation.cpp

```diff
@@ -460,23 +460,15 @@ static void FixupInvocation(CompilerInvocation &Invocation,
     LangOpts.NewAlignOverride = 0;
   }
 
-  // Check whether the SYCL arguments are valid. -fsycl is only accepted in -cc1
-  // if it is passed in conjunction with -fsycl-is-device or -fsycl-is-host. If
-  // -fsycl is not passed but one of those two options is, automatically set it.
-  if (LangOpts.SYCL && !LangOpts.SYCLIsDevice && !LangOpts.SYCLIsHost)
-    Diags.Report(diag::err_drv_argument_only_allowed_with_str)
-        << "-fsycl"
-        << "'-fsycl-is-device' or '-fscyl-is-host'";
-  LangOpts.SYCL = LangOpts.SYCLIsDevice || LangOpts.SYCLIsHost;
-
   // Prevent the user from specifying both -fsycl-is-device and -fsycl-is-host.
   if (LangOpts.SYCLIsDevice && LangOpts.SYCLIsHost)
     Diags.Report(diag::err_drv_argument_not_allowed_with) << "-fsycl-is-device"
                                                           << "-fsycl-is-host";
 
   // Prevent the user from passing -sycl-std= without enabling SYCLIsDevice or
   // SYCLIsHost.
-  if (Args.hasArg(OPT_sycl_std_EQ) && !LangOpts.SYCL) {
+  if (Args.hasArg(OPT_sycl_std_EQ) && !LangOpts.SYCLIsDevice &&
+      !LangOpts.SYCLIsHost) {
     Diags.Report(diag::err_drv_argument_only_allowed_with_str)
         << "-sycl-std="
         << "'-fsycl-is-device' or '-fscyl-is-host'";
```

Comment:
Small preference for this to still say -fsycl' I think, since that is likely what the user is doing.

Reply:
The only downside I see to that is that we'll warn on: `clang -cc1 -sycl-std=2020 foo.cpp` by telling the user to pass `-fsycl` only for them to get a new diagnostic about `-fsycl` being an unknown option.

Perhaps a better approach would be for me to notice when the user does `clang -sycl-std=2020 foo.cpp` without specifying `-fsycl` and diagnose that from the driver side?

Reply:
That is probably optimal now that you mention it.

That said, users aren't typically expected to call our compiler with `-cc1` , right?  It is basically just an implementation detail.

Reply:
Neato, we test that `-sycl-std=` without `-fsycl` has a particular behavior rather than error out. See clang/test/Driver/sycl.c: `// RUN: %clang -### -sycl-std=2017 %s 2>&1 | FileCheck %s --check-prefix=NOT_ENABLED`

So should I change that test to expect an error, or should I assume that a user is relying on this behavior and remove the -cc1 diagnostic (so you can pass `-cc1 -sycl-std=2020` without specifying `-fsycl-is-host` or `-fsycl-is-device` and it's a noop)?

Reply:
(FWIW, I notice we also test this explicitly in CompilerInvocationTest.cpp for the unittests. So my inclination is to leave this be and not warn about it in -cc1.)